### PR TITLE
Fix < menu > for Android and linux

### PR DESCRIPTION
**Pull request by NedScott**
Fix < menu > for Android and Linux. < menu > on Android and Linux normally defaulted to context menu, but this broke when < menu > was added to the global section of keyboard.xml (nearly all remotes on Android are seen as keyboards). This restores the previous behavior, while still adding a way to bring up the sidebar (which is the action ID "menu", not to be confused with "ContextMenu").

### CODE REVIEW

**Comment by NedScott** [comment]
Ping @xhaggi

**Comment by xhaggi** [comment]
does the android remote does not work if you long press the menu button?

**Comment by jjd-uk** [comment]
I think the point here is that Android users are used to pressing the Menu button and getting the Context menu, so when it's instead started opening the sidebar it's confusing users.

**Comment by stefansaraev** [comment]
+1. I think linux users will also prefer having contextmenu on short press.

**Comment by xhaggi** [comment]
if so please do the same for remote.xml to be consistent here. But IMO menu should open menu regardless of what we had before, if a remote does not support long press this needs to be adjusted in that case, but not in our default keymap.

**Comment by NedScott** [comment]
The sidebar menu is a secondary menu and not the more common menu that uses need. Shortpress on menu should go to the most common menu, and that has always been the context menu. "Menu" the button shouldn't do "menu" the action ID simply because it is called menu. If anything, the Menu Action ID should be renamed because it is really confusing at the moment.

Regardless of what supports long press or not, the default for "menu" is context menu. That's how it has always worked, and that's how it should continue to work. I honestly don't care if a keyboard or remote doesn't support long press, because this PR is just about fixing the default behavior for most users.

**Comment by xhaggi** [comment]
As I said if so please adjust all keymaps which includes remote.xml too and we can merge it in asap :)

**Comment by NedScott** [comment]
remote.xml changed as well

**Comment by topfs2** [comment]
On normal android long press is contextual menu. Menu button brings up activity menu. Unless something is different on android tv? Shouldn't we strive to mimic the applied used interaction found on android.

The argument that old android users will be confused is not valid IMO. Every new user will most likely be confused with this change, and we are purposefully teaching them to use kodi different than every other app on android.

**Comment by topfs2** [comment]
Also haven't we always strived to reduce the usage of the context menu. Making it dominant over the sidebar feels like a step back to me.

**Comment by da-anda** [comment]
@topfs - well, context menu is not longpress of menu button but longpress of select button, isn't it? BUT, Kodi's primary usage are not touch devices but TV boxes, where longpress for context menu is not really common, is it?
At least in my opinion the Kodi experience should be the same across devices. Having long press doing A on a android box one but doing B on everything else (HTPC, OE) would at least annoy me from a users perspective

**Comment by topfs2** [comment]
Yeah its long press select. IMO it feels like a valid approach cross platforms to me. The fact that it scales to touch devices is a plus IMO. Honestly I just want the context menu to die in a fire :)

**Comment by topfs2** [comment]
And I guess it comes down to if we want to behave the same cross platforms or behave as the platform we are on. We the devs who run multiple platforms probably want it to behave the same cross but the people who only run and android TV box might not. Not sure which is moat valid here.

**Comment by NedScott** [comment]
topfs2, I think you are confused about what button this affects on a physical remote. This is the same as the "menu" button on an MCE remote, which is still the context menu. Or, for that matter, right click, two flinger press, etc. We already have long-press and long-click for the context menu on iOS, Android, etc.

**Comment by topfs2** [comment]
I am assuming it's this one http://i.stack.imgur.com/hlDK4.png which on nexus has been superseded by the tripple dots in the corner.

**Comment by NedScott** [comment]
Nope :)

It's like this menu button: http://www.china-rc.cc/images/CS-91820140308123216.jpg

or #6 here: http://kodi.wiki/images/b/b6/Amazon_Fire_TV_remote.jpg

**Comment by NedScott** [comment]
Does Kodi even see those buttons on an Android phone/tablet? If so, does it see them as a keyboard? I only have Android boxes. I have no objections to mapping that specific phone/tablet button differently. I wasn't aware it even did anything in Kodi.

**Comment by topfs2** [comment]
Pretty sure all those are handled the same on android, triggers menu. And yes on phone and tablets with that button. Not nexus line.

The icon is just the same on some devices. Ex http://rescom.uwec.edu/support/devices/images/android-button.jpg

**Comment by xhaggi** [comment]
I like the idea to use long press select to open the context menu in addition to the long press menu option.

File: system/keymaps/remote.xml [hunk 1]
```diff
@@ -56,8 +56,8 @@
       <pageplus>PageUp</pageplus>
       <pageminus>PageDown</pageminus>
       <back>Back</back>
-      <menu>Menu</menu>
-      <menu mod="logpress">PreviousMenu</menu>
+      <menu>ContextMenu</menu>
+      <menu mod="logpress">Menu</menu>
```
Comment (phil65) [inline system/keymaps/remote.xml]: typo.

Comment (NedScott) [inline system/keymaps/remote.xml]: fixed

**Comment by topfs2** [comment]
@NedScott pressing menu on samsung triggers context menu on 15.1 which honestly feel weird to me. Im guesing it doesn't trigger the same path then perhaps. I can't build Android so can't go through what triggers what. Honestly though I still find the menu press triggering context menu really wrong on Android.

**Comment by NedScott** [comment]
@topfs2 

I understand that, but there's nothing I can do about it that doesn't sacrifice the button for the larger remote control userbase. Again, and I can't stress this enough, this menu button is the equivalent as MCE menu buttons or the Apple remote menu button, etc, all of which open the context menu.

Not all remotes work with longpress, so without this PR a lot of Linux and Android users no longer have a functioning context menu button. On the flip side, the sidebar menu is always accessible by going left.

Users are already having problems with this on the forum, so can we please restore the v15 behavior already?

**Comment by topfs2** [comment]
I think your missing my point. I'm saying doesn't it make sense that kodi behaves the same as the other apps found on android TV. As I suspect far more users who owns an android TV will use the others apps and _not_ have another kodi box with Linux and an MCE. Your welcome to prove me wrong on this if you have statistics but it feels valid to asssume that most have only one box.

I have a feeling that this change comes from devs which are the only ones that actually notice if the different platforms doesn't behave the same. For most people they would like them to behave as the rest of the platform rather than the same cross platforms.

**Comment by topfs2** [comment]
I don't have an android tv so can't really try it and my pivos isn't hooked up. But start some other apps and check how they behave when you press that menu button and long press.

Also if we don't support long press I'd much rather see us focus on adding that support than doing wierd keymaps ;)

**Comment by NedScott** [comment]
Kodi doesn't behave the same as other android apps. This is the perspective of a user. I don't know any code, and I use the context menu daily in viewing, managing, etc. 

It doesn't matter how many boxes a user has, because restores changes v14's behavior. Our android users are the ones that are getting confused. Having menu open the context menu is essential to remotes with only a few buttons, and even for ones with a lot of buttons but no dedicated management, watched, etc buttons. This is what our users are, in fact, using.

This isn't a weird Keymap. This is reverting a recent change that was done, so that it works as it was designed, and for good reason. 

Having to defend this usage is crazy. It's like being told no one needs a "back" button. This is getting absurd. Someone, please press the damn button.

**Comment by NedScott** [comment]
I'm sorry for that little burst, but I feel like I've stepped into bizarro world here. This is standard remote behavior, topfs2. There has to be some misunderstanding going on, because I can't imagine anyone objecting to this PR. If that is the case, and you don't have access to an HTPC with a remote to see what I'm saying, then please trust me when I say this makes sense and is best for our users.

**Comment by stefansaraev** [comment]
too much bikesheding here. for so simple thing. do we have any chance to get this in. 

@MartijnKaijser

we do support longpress @topfs2 but some remotes dont. because they dont do "repeat". hardware limitations. bla bla. see above.

imo, we should restore the v15 behaviour, and whoever wants to have the behaviour changed, can send a PR and have it discussed. why the hell would we want to piss off any early v16 testers ?

**Comment by FernetMenta** [comment]
+1 for what @stefansaraev said

**Comment by topfs2** [comment]
Merge or whatever I don't care. Just giving my user experience thoughts as I work with that daily.

Nedscott try other apps and check what is standard. As I said I don't have android TV up. I'm just saying we are not doing the standard thing in normal android not on isengard atleast. Menu button does not pop up context menu on other apps on phones, and I'd expect kodi to behave the same as the other apps on the phone when I'm using it on the phone. Same for TV.

And Stefan the support is there on android no?

**Comment by topfs2** [comment]
Also I'm not saying you don't need a back button. Not sure how you got that. I'm saying if all other apps use X-button for back why do we use Y-button.

**Comment by da-anda** [comment]
@topfs2 - problem is, fixing it for Android will break the well known behavior for all other platforms and remotes. So for now we should merge this PR and work on a solution to do platform specific keymaps.

As for longpress - it's there on every platform IIRC, it at least partly works for me on my fireTV (partly means that at least with the fireTV remote a mixture between regular and longpress is being fired in some sort of burst mode, and once you release the key it's a hit and miss if you end up in context menu or the side menu)

**Comment by stefansaraev** [comment]
@da-anda your hit and miss behaviour was a bug. already fixed. not related to this PR.

**Comment by da-anda** [comment]
@stefansaraev good to hear it's fixed - will update to latest nightly. And I know it's not related, was just writing the longpress stuff because topfs2 asked if it's available on Android as well, and it ofc is.

**Comment by NedScott** [comment]
>I'm just saying we are not doing the standard thing in normal android not on isengard atleast. Menu button does not pop up context menu on other apps on phones, and I'd expect kodi to behave the same as the other apps on the phone when I'm using it on the phone.

We are not a phone application. The two UIs can't even compare. The physical menu button on a phone doesn't open a context menu because it is a *touch interface*. There's no directional keys on a phone. You use your finger on a phone. If you could touch your TV in your living room, from 10 feet away, then you would have a point. We cannot design our physical remote controls to be like a touch interface. That's madness.

**Comment by topfs2** [comment]
@NedScott I have said 14 times that I dont have an Android TV to test on, and asked you to test other applications on that how they behave. I'm only comparing to phones as that what I have. And I'm guessing many run it on tablets so it is still somewhat valid. I never said to design it against touch remotes, I'm saying let it behave as the rest of the platform does. What does the menu button _normally_ trigger on Android TV. If its context menu its all fine and dandy, if its the Android Activity menu (which I'm suspecting) then we aren't behaving as other apps. This is _all_ I have ever stated. Do with that what you will

**Comment by NedScott** [comment]
To be blunt, no, I don't think we should strive to be like Android. That's one of the best things about Kodi is that it doesn't work like most Android apps.

Considering the ever changing nature of Android TV (hello, Google TV), I would not look to that as a leader of good UI design. They have several generations of failed products. One of the reasons people like Nvidia wanted us on their hardware was because we actually have a good UI.

As for what other apps to in an Android TV or other Android set-top-box does, that's really besides the point. No other apps do what we do or have the complexity of Kodi. Most of them use "menu" to open up a general settings window, or some kind of settings/quit/notification window. It's almost like an afterthought for those apps. There's not a lot of consistency with what happens between games or other video apps. None of them have any long-press actions like we do now, either. 

IMO, Android be dammed. We are not an Android app. We are Kodi, which has been ported to Android. We are cross platform, first and foremost. The idea that we should fuck it up for Linux and more just to conform to broken Android standards is really strange to me. Imagine if you suggested this in reference to an iOS UI standard? You'd be strung up on a pole by the rest of the team.

The bottom line is, are users surprised when they press the menu button? Nope. They haven't been surprised since our very first Android version. It has always been like this, they like it, it works, it's logical, and it's essential for normal Kodi usage. Without it we have no way to get to the context menu on remotes with keyrepeat limitations.

I know you're not bikeshedding. I have no doubt that someone will eventually press the button and get this merged. I'm just really confused and surprised by your point of view on this.

**Comment by NedScott** [comment]
Sorry for being so combative over this. I know you're just giving your perspective. I tend to go into debate mode easily, and I don't notice how aggressive I've been until I come back and read what I've said. It's nothing personal and just a defect of my own person ;)

Consider my rants on this the same as someone trying to argue Picard vs Kirk. It's easy for someone like myself to rant in detail about silly topics, but it's not in anger or anything like that. If it happens again, just distract me with a picture of boobs, and I'll probably lose interest in debating anything ;)

**Comment by topfs2** [comment]
"The idea that we should fuck it up for Linux and more just to conform to broken Android standards is really strange to me. Imagine if you suggested this in reference to an iOS UI standard? You'd be strung up on a pole by the rest of the team."

I never said touch the other platforms. I suggested be platform specific. So package differently depending on platform so that the keymaps make sense for each platform. It is starting to feel like you haven't read my comments at all, or I'm just bad at explaining. Anyways, I'm out.

**Comment by da-anda** [comment]
@topfs2 - we currently can't have Android exclusive keymaps. Would be nice, but not possible ATM. And for remotes that don't support longpress, triggering the context menu on normal press is currently by far more important than toggling the side blade menu. I totally get your point and am with you, that we have to rethink our keymaps and decide if we want to add platform specific mappings, but let's not do that in this PR.
I also share your vision in regards of client/server, decoupling GUI and have the possibility for native UIs etc, where we definetly need platform specific keymaps, but we're not at this point yet. ATM Kodi looks and behaves the same on every platform and a user shouldn't have to care about the platform Kodi is running on when picking up a remote.

**Comment by topfs2** [comment]
Honestly its not impossible to have platform exclusive keymaps :P You could carry a patch for packaging (OpenELEC does this), a branch for platform, alter the build system to fetch keymaps from specific folders depending during build or even most crude add ifdef in the code to fetch from folders.

Neither of these are overly hard stuff to do if its wanted.

**Comment by NedScott** [comment]
That would only work if we could detect if something was an Android phone vs an Android TV/set-top-box. That's nearly impossible to do, since so many Android set-top-boxes identify as tablets. You want the phone button to behave differently, not the remote button. Even on Android, the menu button on a remote control should open the context menu. Not doing so would be disastrous for Android users. It's safe for phone/tablet users because they will always have long-press screen touch for context menu, and they don't use remote controls.

**Comment by da-anda** [comment]
@topfs2 - it's ofc technically possible, but requires a dev with according knowledge :)
